prober/derp.go

```diff
@@ -577,6 +602,246 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
+	destinationAddr := ifAddr.Next()
+
+	derpProbeBandwidthTUNMu.Lock()
+	defer derpProbeBandwidthTUNMu.Unlock()
+
+	// Temporarily set up a TUN device with which to simulate a real client TCP connection
+	// tunneling over DERP. Use 1280 as our MTU as this is the minimum safe MTU used by
+	// Tailscale.
+	dev, err := tun.CreateTUN("utun9_derpprobe", 1280)
+	if err != nil {
+		return fmt.Errorf("failed to create TUN device: %w", err)
+	}
+	defer func() {
+		if err := dev.Close(); err != nil {
+			log.Printf("failed to close TUN device: %s", err)
+		}
+	}()
+	mtu, err := dev.MTU()
+	if err != nil {
+		return fmt.Errorf("failed to get TUN MTU: %w", err)
+	}
+
+	name, err := dev.Name()
+	if err != nil {
+		return fmt.Errorf("failed to get device name: %w", err)
+	}
+
+	err = configureTUN(prefix, name)
+	if err != nil {
+		return fmt.Errorf("failed to configure tun: %w", err)
+	}
+
+	// Depending on platform, we need some space for headers at the beginning
+	// of TUN packets. The below constant is more than enough space for any
+	// platform that this might run on.
+	tunStartOffset := device.MessageTransportHeaderSize
+
+	// Read packets from TUN device and forward all IPv4 packets destined to
+	// our listener's address via DERP.
+	tunReadC := make(chan error, 1)
+	go func() {
+		numBufs := wgconn.IdealBatchSize
+		bufs := make([][]byte, 0, numBufs)
+		sizes := make([]int, numBufs)
+		for range numBufs {
+			bufs = append(bufs, make([]byte, mtu+tunStartOffset))
+		}
+
+		src, dst := make([]byte, 4), make([]byte, 4)
+		for {
+			n, err := dev.Read(bufs, sizes, tunStartOffset)
+			if err != nil {
+				return
+			}
+
+			for i := range n {
+				pkt := bufs[i][tunStartOffset : sizes[i]+tunStartOffset]
+				// Skip everything except valid IPv4 packets
+				if len(pkt) < 20 {
+					// Doesn't even have a full IPv4 header
+					continue
+				}
+				if pkt[0]>>4 != 4 {
+					// Not IPv4
+					continue
+				}
+
+				destinationAddrBytes := destinationAddr.As4()
+				if !bytes.Equal(pkt[16:20], destinationAddrBytes[:]) {
+					// Unexpected dst address
+					continue
+				}
+
+				copy(src, pkt[12:16])
+				copy(dst, pkt[16:20])
+				copy(pkt[12:16], dst)
+				copy(pkt[16:20], src)
+
+				if err := fromc.Send(toc.SelfPublicKey(), pkt); err != nil {
+					tunReadC <- err
+					return
+				}
+			}
+		}
+	}()
+
+	// Receive the packets.
+	recvC := make(chan error, 1)
+	go func() {
+		buf := make([]byte, mtu+tunStartOffset)
+		for {
+			m, err := toc.Recv()
+			if err != nil {
+				recvC <- fmt.Errorf("failed to receive: %w", err)
+				return
+			}
+			switch v := m.(type) {
+			case derp.ReceivedPacket:
+				if v.Source != fromc.SelfPublicKey() {
+					recvC <- fmt.Errorf("got data packet from unexpected source, %v", v.Source)
+					return
+				}
+				pkt := v.Data
+				copy(buf[tunStartOffset:], pkt)
+				if _, err := dev.Write([][]byte{buf[:len(pkt)+tunStartOffset]}, tunStartOffset); err != nil {
```

Comment:
This might alloc the slice for every write. You can make it once outside the for loop with a length of one instead.